Code/GraphMol/MarvinParse/MarvinDefs.cpp

```diff
@@ -0,0 +1,2978 @@
+              molIDInSgroups)) {
+        sgroupsMolIdsDone.push_back(childSgroup->molID);
+        childSgroup->prepSgroupsForRDKit();
+        allDone = false;
+        break;  // have to start the loop over - we might have changed the
+                // vector
+      }
+    }
+  }
+  // if this is top level, quit
+
+  if (this->parent == nullptr) {
+    return;
+  }
+
+  // check to see if we can move this sgroup up a level. We can if the parent
+  // is not the root of the tree, and its parent is passive (it does not
+  // change the atoms of the parent when it is processed or expanded, or if
+  // all of its atoms do not belong to the the non-passive parent
+
+  std::string parentRole = this->parent->role();
+  std::string childRole = this->role();
+
+  if (parentRole != "MarvinMol") {
+    // check this one
+
+    if (this->parent->isPassiveRoleForExpansion()) {
+      // move the child up
+
+      promoteChild(this);
+      return;
+    }
+
+    // check to see if all of its atomRefs of the child are in the parent.
+    //  there are 3 possibilities:
+    //  1) all atoms are in the parent - do NOT move the child - it belongs
+    //  under the parent 2) NO atoms are in the parent - move the child up 3)
+    //  SOME atoms are in the parent - throw an error
+
+    auto result = this->isSgroupInSetOfAtoms(this->parent->atoms);
+    switch (result) {
+      case SgroupNotInAtomSet:
+        promoteChild(this);
+        return;
+
+      case SgroupInAtomSet:
+        // do noting here
+        break;
+
+      case SgroupBothInAndNotInAtomSet:
+        throw FileParseException(
+            "Child sGroup has atoms both in the parent and NOT in the parent");
+    }
+  }
+
+  // made to here, so the child belongs in the parent.
+
+  if (childRole == "SuperatomSgroup") {
+    ((MarvinSuperatomSgroup *)this)->convertFromOneSuperAtom();
+  } else if (childRole == "MulticenterSgroup") {
+    ((MarvinMulticenterSgroup *)this)->processOneMulticenterSgroup();
+  } else if (childRole == "MultipleSgroup") {
+    ((MarvinMultipleSgroup *)this)->expandOneMultipleSgroup();
+  }
+
+  return;
+}
+
+MarvinMolBase *MarvinSuperatomSgroupExpanded::convertToOneSuperAtom() {
+  // the mol-style super atoms are significatnly different than the Marvin
+  // super atoms.  The mol style has all atoms and bonds in the main mol, and
+  // parameter lines that indicate the name of the super atom and the atom
+  // indices affected.
+  //
+  // The Marvin has a dummy atom with the super atom name in the main
+  // molecule, and a separate sub-mol with the atoms and bonds of the
+  // superatom.  It also has a separate record called attachmentPoint that
+  // atom in the super atom sub=mol that is replaces the dummy atom, and also
+  // a bond pointer and bond order in case the bond order changes.
+  //
+  //  Takes information from a MarvinSuperatomSgroupExpanded and converts the
+  //  MarvinSuperatomSgroup structure
+
+  MarvinMolBase *actualParent = this->parent;
```

Comment:
```suggestion
  PRECONDITION(this->parent,"invalid parent");
  MarvinMolBase *actualParent = this->parent;
```

Reply:
The precondition would throw an error if the pointer is null (an unepected error).   Even without the PRECONDITION, the next line would throw an error. I am not sure this adds anything.

Reply:
In our view a segmentation fault (= uncatchable program crash) is not "throwing an error", It's clearly not always possible, but adding checks in obvious places is a very low-cost defensive measure

Reply:
I added a procedure to find the actual parent that does all the safety checks , and is called from all the places that need it.